flambe/task/training.py

```diff
@@ -0,0 +1,447 @@
+import warnings
+from abc import abstractmethod
+from typing import Dict, List, Optional, Any, Iterable, TypeVar  # noqa: F401
+
+import torch
+import torch.nn as nn
+from torch.optim.optimizer import Optimizer
+from torch.optim.lr_scheduler import _LRScheduler, ReduceLROnPlateau
+from torch.nn.utils.clip_grad import clip_grad_norm_, clip_grad_value_
+
+from flambe.logging import log
+from flambe.learn.task import Task
+
+
+Batch = TypeVar('Batch', bound=Any)
+
+
+class Training(Task):
+    """A PyTorch training task template.
+
+    Implement this interface to use run training and evaluation of
+    a model in PyTorch. This interface requires implementing 3
+    constructors and 3 methods. The interface are designed in a way
+    that the internal repesentation of input examples, and of a batch
+    can be completely generics:
+
+    - ``build_model``: constuct the nn.Module to train. The model
+        is passed to the other methods to execute computation.
+    - ``build_optimizers``: given the model, create the optimizers
+        to use to optimize the model. Should be returned as a dict.
+    - ``build_schedulers``: given the optimizers, create the learning
+        rate schedulers. Should be returns as a dict.
+    - ``sampler``: takes an iterable of example data points and returns
+        an iterable of batches, where a batch can be any object.
+    - ``train_step``: takes the model, and a batch and returns the
+        loss per sample as a tensor of size B (i.e batch size)
+    - ``eval_step``: takes an iterable over batches and returns the
+        metric to use for validation. Higher should indicate better.
+
+    """
+
+    def __init__(self,
+                 max_epoch: float = 1.0,
+                 max_iter: Optional[int] = None,
+                 gradient_accumulation: int = 1,
+                 max_grad_norm: Optional[float] = None,
+                 max_grad_abs_val: Optional[float] = None,
+                 validation_freq: float = 0.01,
+                 eval_only: bool = False,
+                 eval_on_start: List[str] = ['val'],
```

Comment:
These three variables are unused in this file?

Reply:
i believe they are used